lib/Parse/ParseDecl.cpp

```diff
@@ -2601,29 +2623,15 @@ bool Parser::parseNewDeclAttribute(DeclAttributes &Attributes, SourceLoc AtLoc,
   }
   
   case DAK_SwiftNativeObjCRuntimeBase: {
-    if (!consumeIf(tok::l_paren)) {
-      diagnose(Loc, diag::attr_expected_lparen, AttrName,
-               DeclAttribute::isDeclModifier(DK));
-      return false;
-    }
-
-    if (Tok.isNot(tok::identifier)) {
-      diagnose(Loc, diag::swift_native_objc_runtime_base_must_be_identifier);
+    SourceRange range;
+    auto name = parseSingleAttrOption<Identifier>
+                      (*this, Loc, range, AttrName, DK)
+              .whenAny(diag::swift_native_objc_runtime_base_must_be_identifier)
+              .diagnoseWhenOmitted();
```

Comment:
I feel strange `whenAny()` receives a diagnostic. Couldn't it be like
```cpp
  auto name = parseSingleAttrOption<Identifier>
                       (*this, Loc, range, AttrName, DK)
               .diagnoseWhenOmitted(diag::swift_native_objc_runtime_base_must_be_identifier);
```
Or 
```cpp
  auto name = parseSingleAttrOption<Identifier>(
      *this, Loc, range, AttrName, DK,
      diag::swift_native_objc_runtime_base_must_be_identifier)
    .diagnoseWhenOmitted();
```

Reply:
The problem with the first option is that, when you're using `when()`, we currently capture one of the value strings and use it to generate a diagnostic for `diagnoseWhenOmitted()` to use. If `diagnoseWhenOmitted()` were to take a diagnostic, then the attributes that use `when()` would lose this convenience. For instance, this:

```cpp
    auto optMode = parseSingleAttrOption<OptimizationMode>
                            (*this, Loc, AttrRange, AttrName, DK)
                       .when("speed", OptimizationMode::ForSpeed)
                       .when("size", OptimizationMode::ForSize)
                       .when("none", OptimizationMode::NoOptimization)
                       .diagnoseWhenOmitted();
```

Would have to become this:

```cpp
    auto optMode = parseSingleAttrOption<OptimizationMode>
                            (*this, Loc, AttrRange, AttrName, DK)
                       .when("speed", OptimizationMode::ForSpeed)
                       .when("size", OptimizationMode::ForSize)
                       .when("none", OptimizationMode::NoOptimization)
                       .diagnoseWhenOmitted(diag::attr_expected_option_such_as, AttrName, "speed");
```

We could do the second, but we'd probably lose some implementation sharing; that would be a shame considering how similar the logic is.

Reply:
Looking at the specific diagnostics we use:

```
"@_swift_native_objc_runtime_base class name must be an identifier"
"@_objcRuntimeName name must be an identifier"
"@_projectedValueProperty name must be an identifier"
```

Perhaps we should just standardize them so they can all be handled by a single diagnostic—something like:

```
"@%0 argument must be an identifier"
```

We'll lose a bit of specificity (e.g. "name" vs. "class name"), but with that, we could just write `whenAny()` for these without changing the `diagnoseWhenOmitted()` lines.

Reply:
```
  auto optMode = parseSingleAttrOption<OptimizationMode>
                            (*this, Loc, AttrRange, AttrName, DK)
                       .when("speed", OptimizationMode::ForSpeed)
                       .when("size", OptimizationMode::ForSize)
                       .when("none", OptimizationMode::NoOptimization)
                       .diagnoseWhenOmitted(diag::attr_expected_option_such_as, AttrName, "speed");
```
TBH, I like this better than the current form. It's more explicit.

But anyway I'm not a fan of using `AttrOptionSwitch` here, because it's _not_ switching at all. Even if we loose some implementation sharing, `Identifier` version of this logic is probably reasonably trivial, and I believe it would be easier to read.